orchagent/lagids.lua

```diff
@@ -37,24 +37,41 @@ if op == "add" then
         end
         -- proposed lag id is different than that in database OR
         -- the portchannel does not exist in the database
-        -- If proposed lagid is available, return the same proposed lag id
-        if redis.call("sismember", "SYSTEM_LAG_ID_SET", tostring(plagid)) == 0 then
-            redis.call("sadd", "SYSTEM_LAG_ID_SET", tostring(plagid))
-            redis.call("srem", "SYSTEM_LAG_ID_SET", tostring(dblagid))
-            redis.call("hset", "SYSTEM_LAG_ID_TABLE", pcname, tostring(plagid))
-            return plagid
+        -- If proposed lagid is not available, lpop the first availabe ID
+        local index = redis.call("lpos", "SYSTEM_LAG_IDS_FREE_LIST", tostring(plagid))
+        if index == false then
+            local lagid = redis.call("lpop", "SYSTEM_LAG_IDS_FREE_LIST")
+            redis.call("hset", "SYSTEM_LAG_ID_TABLE", pcname, lagid)
+            redis.call("sadd", "SYSTEM_LAG_ID_SET", lagid)
+            if dblagid then
+               redis.call("srem", "SYSTEM_LAG_ID_SET", tostring(dblagid))
+               if redis.call("lpos", "SYSTEM_LAG_IDS_FREE_LIST", tostring(dblagid)) == false then
+                  redis.call("rpush", "SYSTEM_LAG_IDS_FREE_LIST", tostring(dblagid))
+               end
+            end
+            return tonumber(lagid)
+        else
+           redis.call("lrem", "SYSTEM_LAG_IDS_FREE_LIST", 1, tostring(plagid))
+           redis.call("hset", "SYSTEM_LAG_ID_TABLE", pcname, tostring(plagid))
+           redis.call("sadd", "SYSTEM_LAG_ID_SET", tostring(plagid))
+           if dblagid then
+              redis.call("srem", "SYSTEM_LAG_ID_SET", tostring(dblagid))
+              if redis.call("lpos", "SYSTEM_LAG_IDS_FREE_LIST", tostring(dblagid)) == false then
+                 redis.call("rpush", "SYSTEM_LAG_IDS_FREE_LIST", tostring(dblagid))
+              end
+           end
+           return plagid
         end
-    end
-
-    local lagid = lagid_start
-    while lagid <= lagid_end do
-        if redis.call("sismember", "SYSTEM_LAG_ID_SET", tostring(lagid)) == 0 then
-            redis.call("sadd", "SYSTEM_LAG_ID_SET", tostring(lagid))
-            redis.call("srem", "SYSTEM_LAG_ID_SET", tostring(dblagid))
-            redis.call("hset", "SYSTEM_LAG_ID_TABLE", pcname, tostring(lagid))
-            return lagid
+    else
+        local lagid = redis.call("lpop", "SYSTEM_LAG_IDS_FREE_LIST")
+        redis.call("hset", "SYSTEM_LAG_ID_TABLE", pcname, lagid)
+        redis.call("sadd", "SYSTEM_LAG_ID_SET", lagid)
+        if dblagid then
```

Comment:
> why use dblag here?

This is just handling the out of sync cases.  As normal operation, this should not happens.
In the old method, when there is a valid dblagid number (it is from a broken entry in SYSTEM_LAG_ID_TABLE), if a new lagId is assigned, the dblagid needs to be removed from the SYSTEM_LAG_ID_SET. 
With the new modification, we need to handle it as the same -- if it is valid lag_id, we need to it to the free_list.

Reply:
we should do 
```
redis.call("srem", "SYSTEM_LAG_ID_SET", tostring(dblagid))
```
here too ?

Reply:
Yes.  In case of the old lagId (dblagid) still in the SYSTEM_LAG_ID_SET in a broken scenario, we just remove it.  Anyway,  this SYSTEM_LAG_ID_SET is just keeped for debugging purposes.  We don't use it for any functionality.  SYSTEM_LAG_ID_SET and SYSTM_LAG_IDS_FREE_LIST are mutual exclusive.